### DISCUSSION TITLE
Convert SVG to 64 GP image - unable to extend cache

### DISCUSSION THREAD

**Discussion by ttshaw1** [ImageMagick/ImageMagick#3511] [Help]
Hi,

I'm trying to convert an SVG file to a monochrome 250,000x250,000 pixel BMP. I get an error message 

```
magick: unable to extend cache 'out.svg': No space left on device @ error/cache.c/OpenPixelCache/3906.
magick: no decode delegate for this image format `out.svg' @ error/svg.c/RenderRSVGImage/534.
```

According to [this forum post](https://legacy.imagemagick.org/discourse-server/viewtopic.php?t=28763), it's related to the security policy settings. I've tried increasing all the numbers in ImageMagick/policy.xml and ImageMagick/www/source/policy.xml , then deleting all policies. Neither approach has made a difference. Can someone please help me understand what the relevant policy or policies are and what I should change them to?

My system has 64GB of RAM and about 200GB of free space on my C: drive, as well as an E: drive with 931GB free.

**Comment by fmw42** [comment]
What was your exact command line? What is your IM version?

**Reply by ttshaw1** [reply]
The command I'm running is
`magick out.svg -monochrome big_mono_magic.bmp`

IM version ImageMagick 7.0.11-6 Q8 x64 2021-04-03

**Comment by snibgo** [comment]
For large images, see https://www.imagemagick.org/script/architecture.php#tera-pixel

What version of IM? On what platform? (I suppose some version of Windows.)

Please paste the text output from "magick -version".

What SVG rasterizer are you using? If Inkscape, and you don't need any IM processing, it may be better to use Inkscape directly.

You should certainly set your temporary path to a directory on your E: drive.

For the policies: just ensure they are all commented-out. Then paste here the text output from "magick -list resource". EDIT:  Then probably use "-limit memory" and "-limit area" to ensure the pixel cache is forced to disk.

I don't know what memory Inkscape uses. If it is 8 bytes per pixel, then the image needs 500 GB of memory (or, in your case, disk). If your IM is HDRI, that needs 8 bytes/channel/pixel. If IM uses 3 channels, that would be 1500 GB for one image.

**Reply by ttshaw1** [reply]
Thanks for the tera pixel link. I'll read through that.

I'm running ImageMagick 7.0.11-6 Q8 x64 2021-04-03 on Windows 10.

magick -version output:
```
Version: ImageMagick 7.0.11-6 Q8 x64 2021-04-03 https://imagemagick.org
Copyright: (C) 1999-2021 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Visual C++: 192829912
Features: Cipher DPC Modules OpenCL OpenMP(2.0)
Delegates (built-in): bzlib cairo flif freetype gslib heic jng jp2 jpeg jxl lcms lqr lzma openexr pangocairo png ps raw rsvg tiff webp xml zip zlib
```

I don't know what SVG rasterizer I'm using. I'll investigate tomorrow if you think it's important.

magick -list resource output:
```
Resource limits:
  Width: 1.8447EP
  Height: 1.8447EP
  Area: 68.627GP
  List length: unlimited
  Memory: 31.957GiB
  Map: 63.914GiB
  Disk: unlimited
  File: 1536
  Thread: 16
  Throttle: 0
  Time: unlimited
```
Does this tell you whether it's a policy issue or if I'm actually running out of drive space?

I'm running into an error changing the temporary path but will continue tomorrow.

**Comment by snibgo** [comment]
Your command...
```
magick out.svg -monochrome big_mono_magic.bmp
```
... creates a colour image, then converts it to monochrome. So this needs memory to store the rasterized (colour) SVG. I don't know if the conversion to monochrome needs space for another image, or if the conversion is done in-place.

It would help if you could also share your SVG file. If it is small, you can paste it here. Otherwise, you can put it in a zip file, upload the zip somewhere, and paste the link here. (Don't upload the SVG to an imge-sharing website, to prevent the website trying to rasterize a massive image.)

Your IM is Q8, without HDRI. This is good, as each monochrome pixel needs only ony byte. You have 31.9 GiB memory available, so that isn't enough, so IM should use disk for the pixel cache.

If your image is rasterized in colour and you then reduce this to grayscale then you need 3 bytes/pixel, which needs 187 GB of disk, which is too close to your 200 GB available. So you need to use "E:/somedir" for the pixel cache.

Do you really need such a huge image? If not, you can insert "-density 10" or some other number before reading the SVG to get a smaller image.

I suggest you create a directory on E: called "tempim", then try this command:
```
magick -verbose -define registry:temporary-path=E:\tempim -density 10 -limit memory 0 -limit area 0 -limit map 0 -list resource out.svg -monochrome big_mono_magic.bmp
```
Please paste the text output.

If that works, at the low density, try again with "-density 20" etc.

**Reply by dlemstra** [reply]
Another option would be to request the maximum size of the image by adding `-size 1000x1000` instead of density.

**Reply by ttshaw1** [reply]
[SVG attached](https://github.com/ImageMagick/ImageMagick/files/6272758/50mm.zip)

> Your command creates a colour image, then converts it to monochrome. So this needs memory to store the rasterized (colour) SVG. I don't know if the conversion to monochrome needs space for another image, or if the conversion is done in-place.

That's too bad. I was hoping the -monochrome switch would skip over intermediate steps with big color channels and write the bmp as monochrome initially. Do you know of any way that that can be done? I'm hoping to use a simple threshold comparison for each pixel (I tried to set up the svg to be black and white initially, though I think there are grey borders around my shapes), so if the intermediate full color step is there for something like dithering, it's of no value to me.

Unfortunately, I really do need the image to be this large. The .bmp I'm trying to make is the input to a piece of photolithography equipment, with each pixel determining whether or not a ~200nm x 200nm area of resist is exposed. 

Using the command `magick -verbose -define registry:temporary-path=E:\tempim -limit memory 0 -limit area 0 -limit map 0 -monitor -density 10 -list resource 50mm.svg -monochrome -dither None 50mm_10density_no_dither.bmp` (I added -monitor , tried to turn off dithering, and changed the input and output filenames), I get output

```
Resource limits:
  Width: 1.8447EP
  Height: 1.8447EP
  Area: 0P
  List length: unlimited
  Memory: 0B
  Map: 0B
  Disk: unlimited
  File: 1536
  Thread: 16
  Throttle: 0
  Time: unlimited
load image[50mm.svg]: 26417 of 26418, 100% complete
50mm.svg SVG 26418x26418 26418x26418+0+0 8-bit sRGB 41.216MiB 26.531u 0:26.872
contrast-stretch image[50mm.svg]: 26417 of 26418, 100% complete
classify image colors[50mm.svg]: 26417 of 26418, 100% complete
reduce image colors[50mm.svg]: 254 of 255, 100% complete
dither image colors[50mm.svg]: 20922950 of 697910724, 02% complete
```

The dithering step seems to be a major bottleneck, as it's operating on each pixel individually. I'm estimating it would take about 5 hours for the -density 10 image, likely 500 hours for the full image. My attempt to disable dithering seems not to have skipped this step. Did I mess up my command, or is it impossible to skip the dithering calculation?

Thanks a lot for the help. I've made a lot of progress thanks to you.

**Reply by ttshaw1** [reply]
Additional information: when increasing the density to 20, I get the following error:
```
Resource limits:
  Width: 1.8447EP
  Height: 1.8447EP
  Area: 0P
  List length: unlimited
  Memory: 0B
  Map: 0B
  Disk: unlimited
  File: 1536
  Thread: 16
  Throttle: 0
  Time: unlimited
magick: memory allocation failed `50mm.svg' @ error/svg.c/RenderRSVGImage/564.
magick: no images for write '-write' '50mm.bmp' at CLI arg 22 @ error/operation.c/CLINoImageOperator/4893.
```
I can see that the temp location is being used with density 10 and it never seems to get close to filling the E: drive. What could be causing this error?

**Comment by snibgo** [comment]
On dithering: "-dither" is not an operation. It is a [i]setting[/i] that determines whether to apply dithering when reducing colours. See See https://imagemagick.org/script/command-line-options.php#dither . The operation that reduces colours in your command is "-monochrome". Turning off dithering [i]after[/i] reducing colours won't help, it needs to be done [i]before[/i]:
```
... -dither None -monochrome ...
```

**Reply by ttshaw1** [reply]
Nice, that's a massive speedup. Thanks for that.

**Comment by snibgo** [comment]
The OP error was:
```
magick: unable to extend cache 'out.svg': No space left on device @ error/cache.c/OpenPixelCache/3906.
```

The error in your last post is:
```
magick: memory allocation failed `50mm.svg' @ error/svg.c/RenderRSVGImage/564.
```
This is progress: it isn't trying to create a cache in memory. Sadly, it is still running out of memory. Instead of setting the three limits to 0 (zero), try 1GB each.

Note that these limits are not respected by Inkscape. (Your installation seems use IM's internal RSVG. This is good, provided it works.)

For experimenting at small sizes, we can use "-density 0.1". For experimenting at large sizes, I suggest you remove "-monochrome" for now. The main problem is trying to rasterize the large image.

With this command, IM Q16 with HDRI, using Inkscape as the rasterizer:
```
%IMG701010%magick -define registry:temporary-path=H:\temp -limit memory 1GB -limit area 1GB -limit map 1GB 50mm.svg h:\temp\x.mpc
```
... where H: is an external 4 TB drive, with 2 TB free, IM creates two temporary files: 257 GB and 1029 GB. I'll leave it running. It might take all night.

**Reply by ttshaw1** [reply]
Using `magick -verbose -define registry:temporary-path=E:\tempim -limit memory 1GB -limit area 1GB -limit map 1GB -monitor -density 20 -list resource 50mm.svg -dither None -monochrome 50mm.bmp`
I get the following output
```
Resource limits:
  Width: 1.8447EP
  Height: 1.8447EP
  Area: 1000000000P
  List length: unlimited
  Memory: 1000000000B
  Map: 1000000000B
  Disk: unlimited
  File: 1536
  Thread: 16
  Throttle: 0
  Time: unlimited
magick: memory allocation failed `50mm.svg' @ error/svg.c/RenderRSVGImage/564.
magick: no images for write '-write' '50mm.bmp' at CLI arg 22 @ error/operation.c/CLINoImageOperator/4893.
```
After listing the resources, an 11GB file is created in my temp folder. IM then hangs for about 30 seconds before outputting the error message lines. The same is true if I remove the `-dither None -monochrome`. Increasing the density to 30 increases the temp file size to 25GB and the hang time to 2.5 minutes.

Running `magick -define registry:temporary-path=E:\tempim -limit memory 1GB -limit area 1GB -limit map 1GB 50mm.svg E:\tempim\x.mpc` creates two 251GB files. There's been no command line output as of yet. I'm expecting an error after some long time due to the time increasing with image size mentioned in the previous paragraph. 

I'll be interested to see if your task works. Maybe it's an RSVG issue. If so, is the proper way for me to switch to Inkscape to install the command line tool and edit my delegates.xml? Could you please share the relevant line of your delegates.xml if that's the case?

**Comment by snibgo** [comment]
I'm expecting my task to fail because there won't be enough disk space for the output file. But if it gets that far, I'll regard the test as a success -- I would just need another disk for the output.

The test may fail because Inkscape may want to allocate memory for all the pixels. If that were the case, I would expect early failure, and it has been going for 30 minutes now.

But another test worth trying is to run Inkscape directly.

If Inkscape is installed and on your system path, IM should use it, with no change needed in delegates.xml. Inkscape batch commands are documented at https://inkscape.org/doc/inkscape-man.html . One option is "-a" to export part of an image. I don't know if this reduces memory usage. If it does, then the image could be rendered in tiles.

It's too early to suggest whether Inkscape is better for you than RSVG.

>  I was hoping the -monochrome switch would skip over intermediate steps with big color channels and write the bmp as monochrome initially. Do you know of any way that that can be done?

No. I think SVG rasterizing is always done in colour.

**Comment by ttshaw1** [comment]
I've done some more reading and it seems like Microsoft's GDI engine can't handle pixel coordinates greater than 2^15, or 32,768. Setting the density to 12 successfully outputs a 31702px by 31702px image. Setting the density to 13 would generate an image with greater than 32,768 pixels, but instead gives the memory allocation error. I'd start trying to generate the image in Linux at this point, but the tool I need to feed the image into is unfortunately also running on Windows. So it probably wouldn't be able to use any larger image I could generate anyway.

For what it's worth, I tried using Inkscape to convert the svg to png, and after exceeding the same image dimension, the output png file appeared as solid grey with no geometry. I can convert that png to a bmp with magick, but can't open the output bmp to see what it looks like.

**Comment by snibgo** [comment]
> I've done some more reading and it seems like Microsoft's GDI engine can't handle pixel coordinates greater than 2^15, or 32,768.

Where did you read that? And is it relevant?

We can easily create images and files with larger image dimensions, eg:
```
magick -size 1000000x1 gradient: x.bmp
```
>  Setting the density to 13 would generate an image with greater than 32,768 pixels, but instead gives the memory allocation error. 

Oh dear. That suggests a bug somewhere.

**Comment by snibgo** [comment]
Oh dear, yes, with IM v7.0.10-45, this failed after 5 hours:
```
%IMG701010%magick -define registry:temporary-path=H:\temp -limit memory 1GB -limit area 1GB -limit map 1GB 50mm.svg h:\temp\x.mpc

magick: memory allocation failed `50mm.svg' @ error/svg.c/ReadSVGImage/3447.
```
Repeating with "-density 13x1" etc gives the same failure but faster.

I don't have the source code for that version of IM, but it seems to be a failure of cairo_image_surface_create_for_data().

Next test: try the SVG file with "MSVG:" prefix, ie:
```
%IMG701010%magick -verbose -define registry:temporary-path=H:\temp2 -limit memory 1GB -limit area 1GB -limit map 1GB -density 0.1x13 MSVG:s.svg x.png
```

**Comment by ttshaw1** [comment]
> Where did you read that?
Examples: 

- https://stackoverflow.com/questions/2932436/net-gdi-image-size-file-codec-limitations
- https://forums.codeguru.com/showthread.php?407837-GDI-Bitmap-size-limitation
- https://www.codeproject.com/Questions/219830/Large-image-over-32-767-pixels-in-PictureBox

I don't know enough about GDI and ImageMagick to say how relevant this is. Maybe it's unrelated. 

Here's another data point: In an earlier test, I managed to make a BMP that's 32840x10143 from the [attached SVG](https://github.com/ImageMagick/ImageMagick/files/6275131/tiled_snail_array.zip)
 using the python library reportlab, which doesn't do anything helpful for large images like caching to disk. Yet if I try to convert it in ImageMagick I get the memory allocation error. At the same time, I can create a (31701px)^2 image without issue that's three times larger in area. 

So it appears IM can't do a conversion with one dimension exceeding the limit, while other software can. I was able to generate a 33000x33000 bmp `magick -size 33000x33000 gradient: x.bmp`, so it seems like the issue has to do with the conversion from SVG.

I'm running your MSVG: command overnight. We'll see what comes of that.

As a sidenote, do you have any recommendations for robust image viewers for large images such as those we're working with? None of mine will open images larger than (20,000 pixels)^2 or so.

**Comment by snibgo** [comment]
Thanks for the GDI links. They are ten years old, so that limitation may no longer apply.

IM doesn't use GDI, as far as I know.

Viewers for large images? Maybe nip2, https://github.com/libvips/nip2 . This is based on libvips, which was designed for very large images. The nip2 interface is more like a spreadsheeet than a conventional image editor. I haven't used it for years.

My own image editor, "swim", which uses IM, also works. But I don't publish it. It is massively slow when the pixel caches are on an external USB hard disk, eg with 49600x70100 pixels. Maybe performance would be okay using SSD (Solid State Drives).

**Reply by ttshaw1** [reply]
```
magick -verbose -define registry:temporary-path=E:\tempim -limit memory 1GB -limit area 1GB -limit map 1GB -density 0.1x13 MSVG:50mm.svg 50mm_test.png
mvg:E:/tempim/magick--gMivrqhJT8PGIRuo_vMM7TRA50EdZY8=>E:/tempim/magick--gMivrqhJT8PGIRuo_vMM7TRA50EdZY8 MVG 264x34344 264x34344+0+0 8-bit sRGB 47.919MiB 19198.547u 319:54.558
MSVG:50mm.svg=>50mm.svg SVG 264x34344 264x34344+0+0 8-bit sRGB 47.919MiB 0.000u 0:00.005
MSVG:50mm.svg=>50mm_test.png SVG 264x34344 264x34344+0+0 8-bit sRGB 47.919MiB 5.094u 0:01.804
```
This command worked! I'm currently retrying with the full density.

**Comment by snibgo** [comment]
And for me: the MSVG test worked, hoorah!
```
%IMG701010%magick -verbose -define registry:temporary-path=H:\temp2 -limit memory 1GB -limit area 1GB -limit map 1GB -density 0.1x13 MSVG:s.svg x.png
```
The output x.png is 264x34344 pixels. Sadly, it took 16 hours. Ouch. I conclude that a full-size rasterizing would work, given enough time and disk space. But the time isn't practical.

The SVG contains many polygons. They seem to all be rectangles. If so, then "rect" may be much faster than "polygon", because the test for inside/outside a rectangle is much easier than inside/outside a general polygon.

```
inkscape --export-dpi=13 --export-filename i.png s.svg
```
The "-a" option works:
```
inkscape --export-area=0:0:999:999 --export-filename i2.png s.svg
```
My current thinking is that using IM to rasterize that SVG isn't reasonable. Instead, use Inkscape or something else. If the rasterizing is done in tiles, IM can process each tile and append them together.
